### ISSUE TITLE
Implement Convolution Transpose Layer

### ISSUE THREAD

**Issue by adam-r-kowalski** [tensorflow/swift-apis#53]
While a Convolutional layer already is implemented in the existing layers, if we had a transposed version we could start to build up interesting architectures such as GANs and VAEs.

**Comment by tanmayb123** [comment]
I’ll see if I can try to implement these layers later this week. It shouldn’t be _too_ difficult, just use `_TFConv2DBackpropInput` (https://github.com/apple/swift/blob/bcb73a17770da5b891fb8cfbc647c6425bf30e53/stdlib/public/TensorFlow/Gradients.swift#L648).

The implementation in TensorFlow (Python): https://github.com/tensorflow/tensorflow/blob/6612da89516247503f03ef76e974b51a434fb52e/tensorflow/python/ops/nn_ops.py#L1578

**Comment by tanmayb123** [comment]
@rxwei I've got a **_very-very-rough_** version of this layer here: https://gist.github.com/tanmayb123/d1509a0c294a39bd6369cfbf3a6af944

The forward pass works nicely, but when I try to calculate the gradient using `gradient(at:)` I get this error:

```
Fatal error: Conv2DCustomBackpropFilter: input must be 4-dimensional: file /swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2094
Current stack trace:
0    libswiftCore.so                    0x00007f6676eaaf40 _swift_stdlib_reportFatalErrorInFile + 115
1    libswiftCore.so                    0x00007f6676df33dc <unavailable> + 3003356
2    libswiftCore.so                    0x00007f6676df34ce <unavailable> + 3003598
3    libswiftCore.so                    0x00007f6676c3ae12 <unavailable> + 1199634
4    libswiftCore.so                    0x00007f6676dbd6b2 <unavailable> + 2782898
5    libswiftCore.so                    0x00007f6676c3a259 <unavailable> + 1196633
6    libswiftTensorFlow.so              0x00007f66742cead2 <unavailable> + 441042
7    libswiftTensorFlow.so              0x00007f66742cd230 checkOk(_:file:line:) + 491
8    libswiftTensorFlow.so              0x00007f66742f0270 _TFCCheckOk(_:) + 81
9    libswiftTensorFlow.so              0x00007f66742f0260 _swift_tfc_CheckOk + 9
```

I've tried debugging this, but I'm a bit stuck. Do you have any ideas as to how I could fix this?

**Comment by rxwei** [comment]
Gradient is only defined for functions that return a scalar. Yes, a better error message would be helpful.

To test the gradient, sum up the result to a scalar.
```swift
gradient(at: b) { c -> Tensor<Float> in
  return c.applied(to: a, in: Context(learningPhase: .training)).sum()
}
```

**Comment by tanmayb123** [comment]
Ah. I've been pretty confused at that ;)

**Comment by tanmayb123** [comment]
Quick question Richard: `_TFConv2DBackpropInput` is currently at an `internal` protection level. How would I go about making that public? Could we make that an op instead of a gradient (i.e. move that from `Gradients.swift` to `Ops.swift`)?

**Comment by rxwei** [comment]
I'd actually suggest moving the definition of all convolution APIs from the standard library to swift-apis!

**Comment by tanmayb123** [comment]
That's a great idea. Just so I'm clear, can I create two PRs to move these:

1. `batchNormalized` op
1. `_TFConv2DBackpropInput` op
1. `_TFConv2DBackpropFilter` op
1. `convolved2D` op
1. `averagePooled` op
1. `maxPooled` op
1. `Padding` enum

To a new `Ops.swift` file in this repo?

**Comment by rxwei** [comment]
[swift-apis/Sources/DeepLearning/Operators.swift](Sources/DeepLearning/Operators.swift) would be the best place to put them.

**Comment by tanmayb123** [comment]
@rxwei quick question: I'm trying to create a differentiable function that can determine the output shape of the transposed convolution layer. This is what I've got:

```
    @inlinable
    @differentiable
    func transposedConv2DOutputShapeValidPadding(
        filterShape: Tensor<Int32>, strides: Tensor<Int32>) -> Tensor<Int32> {
        let bs = shapeTensor[0]
        let w = (shapeTensor[1] - 1) * strides[1] + filterShape[0]
        let h = (shapeTensor[2] - 1) * strides[2] + filterShape[1]
        let c = shapeTensor[3]
        return Tensor<Int32>([bs, w, h, c])
    }

    @inlinable
    @differentiable
    func transposedConv2DOutputShapeSamePadding(
        filterShape: Tensor<Int32>, strides: Tensor<Int32>) -> Tensor<Int32> {
        let w = shapeTensor[1] * strides[1]
        let h = shapeTensor[2] * strides[2]
        return Tensor<Int32>([shapeTensor[0], w, h, shapeTensor[3]])
    }
```

Of course, this isn't differentiable:

```
error: <Cell 37>:142:6: error: function is not differentiable
    @differentiable
    ~^~~~~~~~~~~~~~

<Cell 37>:143:10: note: when differentiating this function definition
    func transposedConv2DOutputShapeSamePadding(
         ^

<Cell 37>:147:31: note: cannot differentiate through a non-differentiable argument; do you want to add '.withoutDerivative()'?
        return Tensor<Int32>([shapeTensor[0], w, h, shapeTensor[3]])
                              ^

error: <Cell 37>:142:6: error: function is not differentiable
    @differentiable
    ~^~~~~~~~~~~~~~

<Cell 37>:143:10: note: when differentiating this function definition
    func transposedConv2DOutputShapeSamePadding(
         ^

<Cell 37>:147:31: note: expression is not differentiable
        return Tensor<Int32>([shapeTensor[0], w, h, shapeTensor[3]])
                              ^

error: <Cell 37>:131:6: error: function is not differentiable
    @differentiable
    ~^~~~~~~~~~~~~~

<Cell 37>:132:10: note: when differentiating this function definition
    func transposedConv2DOutputShapeValidPadding(
         ^

<Cell 37>:138:31: note: cannot differentiate through a non-differentiable argument; do you want to add '.withoutDerivative()'?
        return Tensor<Int32>([bs, w, h, c])
                              ^

error: <Cell 37>:131:6: error: function is not differentiable
    @differentiable
    ~^~~~~~~~~~~~~~

<Cell 37>:132:10: note: when differentiating this function definition
    func transposedConv2DOutputShapeValidPadding(
         ^

<Cell 37>:138:31: note: expression is not differentiable
        return Tensor<Int32>([bs, w, h, c])
```

Any tips as to how I can make this differentiable? `_TFConv2DBackpropInput` requires knowing its expected output shape.

**Comment by tanmayb123** [comment]
Also - separate question - is casting tensors to different values going to be differentiable? Will I be able to do something like this:

```
@differentiable
func simpleadder(abc: Tensor<Int32>) -> Tensor<Float> {
    return Tensor<Float>(abc) + Tensor<Float>(1)
}
```

**Comment by tanmayb123** [comment]
I've implemented this layer in #64.

**Comment by sjaz24** [comment]
Trying to use the TransposedConv2D layer but getting strange error.  Below is a small example that when I run either in Colab or locally with a recent build, I get the following error. Any ideas??

`Fatal error: Incompatible shapes: [1,4,4,256] vs. [128]: file /swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2123
Current stack trace:
0    libswiftCore.so                    0x00007f06347314a0 _swift_stdlib_reportFatalErrorInFile + 115
1    libswiftCore.so                    0x00007f063467930c <unavailable> + 3035916
2    libswiftCore.so                    0x00007f06346793fe <unavailable> + 3036158
3    libswiftCore.so                    0x00007f06344c06c2 <unavailable> + 1230530
4    libswiftCore.so                    0x00007f0634646292 <unavailable> + 2826898
5    libswiftCore.so                    0x00007f06344bfba9 <unavailable> + 1227689
6    libswiftTensorFlow.so              0x00007f0631aae572 <unavailable> + 599410
7    libswiftTensorFlow.so              0x00007f0631aaccc0 checkOk(_:file:line:) + 508
8    libswiftTensorFlow.so              0x00007f0631ad1ad0 _TFCCheckOk(_:) + 81
9    libswiftTensorFlow.so              0x00007f0631ad1ac0 _swift_tfc_CheckOk + 9
Current stack trace:
	frame #11: 0x00007f0661da2be7 $__lldb_expr68`AD__$s14__lldb_expr_6710TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___primal_src_0_wrt_0(layer=<unavailable>) at <Cell 12>:14:15
	frame #12: 0x00007f0661da3b84 $__lldb_expr68`AD__$s14__lldb_expr_6710TensorFlow0C0VySfGAB16TransposedConv2DVXEfU___vjp_src_0_wrt_0 at <Cell 12>:13:56
	frame #20: 0x00007f0661da19a0 $__lldb_expr68`main at <Cell 12>:13:38`

`import TensorFlow

let tensorShape = TensorShape([4, 4, 256, 64])
let seed = (Int64.random(in: Int64.min..<Int64.max), Int64.random(in: Int64.min..<Int64.max))
let transposedConv2D = TransposedConv2D(
  filter: Tensor(glorotUniform: tensorShape, seed: seed),
  bias: Tensor(zeros: TensorShape([128])),
  activation: identity,
  strides: (1, 1),
  padding: .valid)

let tensor = Tensor<Float>(randomUniform: TensorShape([1,1,1,64]))
let (loss, grads) = transposedConv2D.valueWithGradient { layer -> Tensor<Float> in
    let out = layer(tensor)
    print(out.shape)
    let loss = out.sum()
    print(loss)
    return loss
}`

**Comment by Shashi456** [comment]
@sjaz24 I opened #282 for this, lets shift this conversation there, I have encountered some similar errors, would be happy to talk to you about them.
